### PR TITLE
Dynamic import expressions

### PR DESCRIPTION
**Pull request by coagmano**
Aims to solve #394 

Also took the liberty of including @benjamn's words from [here](https://github.com/meteor/meteor/pull/8327#issuecomment-281428641) and [here](https://github.com/meteor/meteor/pull/8327#issuecomment-280876264) explaining the difference between Meteor's system and other bundling tools.

Happy to take constructive feedback on content, writing style, structure, grammar, etc!

### CODE REVIEW

File: source/packages/dynamic-import.md [hunk 1]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
...
+  import("/libs/yet-another-example.js");
+}
+```
+This whitelist must be imported in both client and server code so the 
+referenced modules are included. Make sure the whitelist is imported from the 
+client and server entry points.
+
+## Difference to other bundling systems
+
+In Meteor's implementaiton, the client has perfect information about which
```
Comment (arggh) [inline source/packages/dynamic-import.md]: Typo in `implementaiton`

File: source/packages/dynamic-import.md [hunk 2]
```diff
@@ -8,7 +8,7 @@ description: Documentation of Meteor's `dynamic-import` package.
 The `dynamic-import` package provides an implementation of
 `Module.prototype.dynamicImport`, an extension of the module runtime which
 powers the [dynamic `import(...)`](https://github.com/tc39/proposal-dynamic-import)
-statement, an up-and-coming (currently stage 3 out of 4) addition to the
+statement, an up-and-coming (ECMA2020) addition to the
 ECMAScript standard.
 
 The dynamic `import(...)` statement is a complimentary method to the static
```
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
The dynamic `import(...)` statement is a complementary method to the static
```

File: source/packages/dynamic-import.md [hunk 3]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
...
+let path = 'example';
+const module = await import(`/libs/${path}.js`);
+```
+You'll get an error like so:
+```js
+Error: Cannot find module '/libs/example.js'
+```
+
+Meteor’s build process builds a graph of all files that are imported or required 
+using static analysis. It then creates exact bundles of the the referenced files
```
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
using static analysis. It then creates exact bundles of the referenced files
```

File: source/packages/dynamic-import.md [hunk 4]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
 > ```js
 > import("another-tool").then(({ default: thatTool }) => thatTool.go());
 > ```
+
+### Using `import()` with dynamic expressions
+
+If you try to import using any computed expression like so:
```
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
If you try to import using any computed expression, such as:
```

File: source/packages/dynamic-import.md [hunk 5]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
...
+
+Meteor’s build process builds a graph of all files that are imported or required 
+using static analysis. It then creates exact bundles of the the referenced files
+and makes them available to the client for `import()`.
+
+Without a complete import statement (static, dynamic or `require`), Meteor won't
+make that module available for `import()`.
+
+The solution to make dynamic expressions work is to create a module white-list 
+that can be read by the build process, but does not actually run like so:
```
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
that can be read by the build process, but does not actually run. For example:
```

File: source/packages/dynamic-import.md [hunk 6]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
...
+```js
+if (false) {
+  import("/libs/example.js");
+  import("/libs/another-example.js");
+  import("/libs/yet-another-example.js");
+}
+```
+This whitelist must be imported in both client and server code so the 
+referenced modules are included. Make sure the whitelist is imported from the 
+client and server entry points.
```
Comment (robfallows) [inline source/packages/dynamic-import.md]: I think you can drop the first sentence. The second says it all.

File: source/packages/dynamic-import.md (hunk 1 above)
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
In Meteor's implementation, the client has perfect information about which
```

File: source/packages/dynamic-import.md [hunk 7]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
...
+response from the server. You might call this strategy **exact code splitting**,
+to differentiate it from bundling.
+
+Moreover, the initial bundle includes the hashes of all available dynamic
+modules, so the client doesn't have to ask the server if it can use a cached
+version of a module, and the same version of the module never needs to be
+downloaded again by the same client. This caching system has all the benefits of
+immutable caching.
+
+Meteor also allows dynamic expressions as the dependency is expressed statically
```
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
Meteor also allows dynamic expressions as long as the dependency is expressed statically
```

File: source/packages/dynamic-import.md [hunk 8]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
...
+somewhere else in your code. This is possible because Meteor's client-side module
+system understands how to resolve dynamic strings at runtime (which is not true
+in webpack or browserify, because they replace module identifier strings with 
+numbers). However, the set of available modules is constrained by the string
```
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
numbers). However, the set of available modules is constrained by the
```

File: source/packages/dynamic-import.md [hunk 9]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
...
+version of a module, and the same version of the module never needs to be
+downloaded again by the same client. This caching system has all the benefits of
+immutable caching.
+
+Meteor also allows dynamic expressions as the dependency is expressed statically
+somewhere else in your code. This is possible because Meteor's client-side module
+system understands how to resolve dynamic strings at runtime (which is not true
+in webpack or browserify, because they replace module identifier strings with 
+numbers). However, the set of available modules is constrained by the string 
+literals that you, the programmer, explicitly decided to allow to be imported.
```
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
whitelist that you, the programmer, explicitly decided to allow to be imported.
```

File: source/packages/dynamic-import.md [hunk 10]
```diff
@@ -61,3 +61,57 @@ async function performTask() {
...
+```
+
+Meteor’s build process builds a graph of all files that are imported or required 
+using static analysis. It then creates exact bundles of the the referenced files
+and makes them available to the client for `import()`.
+
+Without a complete import statement (static, dynamic or `require`), Meteor won't
+make that module available for `import()`.
+
+The solution to make dynamic expressions work is to create a module white-list
```
Comment (robfallows) [inline source/packages/dynamic-import.md]:
```suggestion
The solution to make dynamic expressions work is to create a module "whitelist"
```

File: source/packages/dynamic-import.md (hunk 9 above)
Comment (coagmano) [reply source/packages/dynamic-import.md]: Thanks for the review. This was the only one I haven't added to the new commit, as this part is explaining module splitting in general, which works for all literal imports and not just whitelists. 

I could add another reference to the whitelist here?
Something like this? 
> However, the set of available modules is constrained by the string literals that you, the programmer, explicitly decided to allow to be imported (either directly or in a whitelist).

File: source/packages/dynamic-import.md [hunk 11]
```diff
@@ -49,15 +49,74 @@ async function performTask() {
...
+The `import(...)` `Promise` is resolved with the `exports` of the module.
+If it's necessary to use the "default" export from a module, it will be
+available on the `default` property of the resulting object.  In the above
+examples, this means it will be available as `tool.default`.  It can be
+helpful to use parameter de-structuring to provide additional clarity:
+
+```js
+import("another-tool").then(({ default: thatTool }) => thatTool.go());
+```
+{% endblockquote %}
```
Comment (coagmano) [inline source/packages/dynamic-import.md]: For some reason, having a code block inside a blockquote works fine if it's the end of the file. Adding extra content caused it to break, so I fixed it by changing the markdown blockquote to a hexo-nunjucks one and it works fine now 🤷‍♂️

File: source/packages/dynamic-import.md (hunk 9 above)
Comment (robfallows) [reply source/packages/dynamic-import.md]: Sounds good to me. 👍

**Pull request merged**